---
title: "Identification and analysis of spatial patterns"
author: "Jovan Tanevski"
date: "10.05.2023"
output:
  html_document:
    df_print: paged
---

# Setup

Install Bioconductor and the required packages.

```{r message=FALSE}
library(TENxVisiumData)
library(SpatialExperiment)
library(ggspavis)
library(sctransform)
library(spdep)
library(SPARK)
library(OmnipathR)
library(decoupleR)
library(tidyverse)
library(scran)
```

# Dataset

One of the common formats for storing sptial data in R is `SpatialExperiment`. Familiarize yourself with the details of the storage object and how to access various aspects of the data.

First resource is the vignette of the package `SpatialExperiment`

```{r}
vignette("SpatialExperiment")
```

The vignette is also available [online](https://bioconductor.org/packages/release/bioc/vignettes/SpatialExperiment/inst/doc/SpatialExperiment.html).

More information can be found in this [article](https://academic.oup.com/bioinformatics/article/38/11/3128/6575443) or in the online textbook on ["Orchestrating Spatially Resolved Transcriptomics Analysis with Bioconductor" (OSTA)](https://lmweber.org/OSTA-book/).

The package `TENxVisiumData` allows easy access to a collection of publicly available 10x Visium spatial transcriptomics datasets in `SpatialExperiment` format. You can find a complete list of available datasets by running `?TENxVisiumData`. More information about the data can be found on the [10x website](https://www.10xgenomics.com/resources/datasets).

For starters we will use a sample of a tissue from a human heart. In the following snippet you will find an example of plotting SpatialExperiment data with `ggspavis` . When plotting the abundance of a single gene, the symbol is given as an [Ensembl](www.ensembl.org) id.

A convinient map from Ensembl id to a gene symbol can be found in the `rowData` field of the `SpatialExperiment` object.

There are several ways how to normalize the data and stabilize the variance across genes. Here, after removing duplicate symbols and genes with total count of zero, we use regularized negative binomial regression.

Note also that the spatial information that we are going to use is not the scaled pixel location available from `spatialCoords(spe)` , but rather the true row and column of the Visium spot in the tissue array.

```{r}

spe <- HumanHeart()
plotVisium(spe, spots=FALSE)
# MYH7
plotMolecules(spe, molecule = "ENSG00000092054", palette = c("white", "navy"))
# COL1A1
plotMolecules(spe, molecule = "ENSG00000108821")
data.raw.counts <- counts(spe)

symbols <- rowData(spe)[,"symbol"]
rownames(data.raw.counts) <- symbols

data.raw.counts.filtered <- data.raw.counts[(!duplicated(symbols) & rowSums(data.raw.counts) != 0), ]

data.vst <- vst(data.raw.counts.filtered, verbosity = 1)$y

hvgs <- getTopHVGs(modelGeneVar(data.vst), n = 1000)
  
data.geometry <- int_colData(spe)$spatialData[,c("array_row", "array_col")]


```

# Spatial patterns

The goal of the practical is to identify genes that exhibit spatial patterns. There are many approaches how to do this. The most basic approaches are based on measuring the spatial autocorrelation of a gene. There are several more advanced ways to detect genes with spatial patterns. Here we have an example of using SPARK with extended number of kernels (pattern templates).

## Spatial autocorrelation

Calculate the spatial autocorrelation of the top 1000 highly variable genes.

### Global spatial autocorrelation

Moran's I

$$
I = \frac{N}{W}\frac{\sum_{i=1}^N\sum_{j=1}^Nw_{ij}(x_i - \bar{x})(x_j - \bar{x})}{\sum_{i=1}^N(x_i - \bar{x})^2}
$$
Determine the neighborhood of each spot and the values of *w*. The lower and upper bounds of the distance between the spots considered as neighbors are currently set to 0 and 2. Play around with these values and observe the effect on the results below.

```{r}
neighbors <- dnearneigh(data.geometry %>% as.matrix(), 0, 2)
weights <- nb2listw(neighbors, style = "B")
```

Calculate the global spatial autocorrelation for each of the highly variable genes by Moran's I. This might take a minute. Be patient.
```{r}
moranI <- hvgs %>% 
  map_dfr(~tibble_row(gene = .x, 
                      autoc = moran.test(data.vst[.x,], weights)$estimate[1]), 
          .progress = TRUE)

```



### Local spatial autocorrelation

Geary's C

$$
C = \frac{(N-1)}{2W}\frac{\sum_{i=1}^N\sum_{j=1}^Nw_{ij}(x_i - x_j)^2}{\sum_{i=1}^N(x_i - \bar{x})^2}
$$

Calculate the local spatial autocorrelation for each of the highly variable genes by Geary's C. This might take a minute. Be patient.

```{r}
gearyC <- hvgs %>% 
  map_dfr(~tibble_row(gene = .x, 
                      autoc = geary.test(data.vst[.x,], weights)$estimate[1]), 
          .progress = TRUE)

```

## SPARK

Note that SPARK works with raw counts.

```{r}
spark.patterns <- sparkx(data.raw.counts.filtered, data.geometry, numCores = 4)
```

# Functional annotation of patterns

Once we have quantified in different ways the spatial patterns of expression of different genes, we are interested in the functional annotation of the set of genes with spatial patterns.

```{r}
msigdb <- get_resource("MSigDB")


enrich.spark <- run_ora(1 - spark.patterns$res_mtest %>% select(adjustedPval) %>% 
                   as.matrix(), 
                 msigdb %>% filter(collection == "hallmark") %>% 
                   select(-uniprot) %>% distinct(), 
                 .source="geneset", .target="genesymbol")

enrich.spark %>% arrange(p_value)

```

Can we claim that the some of the hallmark states or processes show spatial patterns? Why not? What type of analysis would support such a claim? 

# Further work

-   Following the SPARK example, enrich the results obtained by calculating Moran's I (and Geary's C). Be careful! The values of Moran's I and Geary's C are on different ranges ([-1, 1] vs [0, Inf]) and have a different meaning. What are the differences in the result of the enrichment? 

-   Try enrichment per individual spark kernel scores (Hint: `spark.patterns$stats` or `spark.patterns$res_stest`).

-   What are the constituent genes of the pathways? Check in the `msigdb` object. Plot some of them with high scores or low p-values to visualize the spatial patterns.

-   Are genes with significant spatial patterns co-localized? Visualize or try to quantify.

-   Try running the same analysis on a cancer sample. Where and what are the differences? 

```{r}

# space for your own work 



```

